File: 0767-reorganize-string/0767-reorganize-string.java
class Solution {
    public String reorganizeString(String s) {
        var cnt = new int[26];
        for(char c : s.toCharArray()) cnt[c-'a']++;
        var pq = new PriorityQueue<int[]>((a, b) -> Integer.compare(b[1], a[1]));
        for(int i=0; i<26; i++){
            if(cnt[i] > 0) pq.offer(new int[] {i+'a', cnt[i]});
        }
        var ans = new StringBuilder();
        while(!pq.isEmpty()){
            var first = pq.poll();
            if(ans.length()==0 || first[0] != ans.charAt(ans.length()-1)){
                ans.append((char)first[0]);
                if(--first[1] > 0) pq.offer(first);
            }
            else{
                if(pq.isEmpty()) return "";
                var second = pq.poll();
                ans.append((char) second[0]);
                if(--second[1] > 0) pq.offer(second);
                pq.offer(first);
            }
        }
        return ans.toString();
    }
}

File: 0767-reorganize-string/0767-reorganize-string.cpp
class Solution {
public:
    string reorganizeString(string s) {
        vector<int> cnt(26, 0);
        for(char c : s) cnt[c-'a']++;
        priority_queue<vector<int>> pq;
        for(int i=0; i<26; i++){
            if(cnt[i] > 0) pq.push(vector<int>{cnt[i], i+'a'});
        }
        string ans;
        while(!pq.empty()){
            auto first = pq.top(); pq.pop();
            if(ans.empty() || first[1] != ans.back()){
                ans += char(first[1]);
                if(--first[0] > 0) pq.push(first);
            }
            else{
                if(pq.empty()) return "";
                auto second = pq.top(); pq.pop();
                ans += char(second[1]);
                if(--second[0] > 0) pq.push(second);
                pq.push(first);
            }
        }
        return ans;
    }
};